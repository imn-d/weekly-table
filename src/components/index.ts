export * from './Background';
export * from './Popup';
export * from './TimeBlockCore';
export * from './CustomTime';
export * from './SchedulerArea';
export * from './SchedulerHeader';
export * from './SchedulerHelper';
export * from './SchedulerLayout';
