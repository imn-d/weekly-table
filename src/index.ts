import {
  BlockColorsProps,
  ScheduleGroup,
  SchedulerColumnsProps,
  SchedulerInputProps,
  schedulerColumns,
  schedulerRows,
} from './common';
import Scheduler from './providers/Scheduler';

export default Scheduler;
export { schedulerColumns, schedulerRows };
export type {
  SchedulerColumnsProps,
  SchedulerInputProps,
  ScheduleGroup,
  BlockColorsProps,
};
