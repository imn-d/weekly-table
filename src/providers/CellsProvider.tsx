import React, {
  FC,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

import { Cell, CellsProps, HOUR_24 } from '../common';
import { useScheduler } from './Scheduler';

const CellsContext = createContext<CellsProps>({} as CellsProps);
/**
 *  Virtual cells provider context hook
 *  @returns [CellsProps]{@link CellsProps}
 */
export const useCells = (): CellsProps => useContext(CellsContext);

/**
 * Virtual cells provider
 */
const CellsProvider: FC = ({ children }) => {
  const {
    parentRef,
    timeframe,
    columns,
    helperWidthProp,
    headerHeightProp,
    bottomHeightProp,
  } = useScheduler();

  /**
   * Main context state, setting by useEffect
   */
  const [cellsProps, setCellsProps] = useState<CellsProps>({} as CellsProps);

  /**
   * Main context updater
   */
  const setCells = useCallback(
    (update: Cell[]) => {
      setCellsProps({ ...cellsProps, cells: update });
    },
    [cellsProps],
  );

  /**
   * Ms count per timeframe
   */
  const msTime = useMemo(() => {
    return timeframe * 60 * 1000;
  }, [timeframe]);

  /**
   * Calculates context values and build virtual cells area
   */
  useEffect(() => {
    const parent = parentRef?.current;
    if (!parent) return;
    const widthStep = (parent.offsetWidth - helperWidthProp) / columns?.length;
    const areaHeight =
      parent.offsetHeight - headerHeightProp - bottomHeightProp;
    const heightStep = areaHeight / (1440 / timeframe);
    const millisPerPixel = HOUR_24 / areaHeight;
    const maxCellIndex = areaHeight / heightStep - 1;
    const cells = new Array<Cell>();
    let widthPos = helperWidthProp;
    for (let i = 0; i < columns?.length; i++) {
      let heightPos = headerHeightProp;
      for (let j = 0; j < areaHeight / heightStep; j++) {
        cells.push({
          isSelected: false,
          column: i,
          row: j,
          position: {
            x: widthPos,
            y: heightPos,
            w: widthPos + widthStep,
            h: heightPos + heightStep,
          },
        } as Cell);
        heightPos += heightStep;
      }
      widthPos += widthStep;
    }

    setCellsProps({
      widthStep,
      heightStep,
      millisPerPixel,
      cells,
      maxCellIndex,
      offsetWidth: parent.offsetWidth,
      offsetHeight: parent.offsetHeight,
    } as CellsProps);
  }, [
    parentRef?.current?.offsetWidth,
    parentRef?.current?.offsetHeight,
    columns?.length,
    bottomHeightProp,
    headerHeightProp,
    helperWidthProp,
    timeframe,
    parentRef,
  ]);

  return (
    <CellsContext.Provider
      value={{
        ...cellsProps,
        setCells: setCells,
        msTime,
      }}
    >
      {children}
    </CellsContext.Provider>
  );
};

export default CellsProvider;
