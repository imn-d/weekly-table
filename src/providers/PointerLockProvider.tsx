import React, {
  FC,
  createContext,
  useCallback,
  useContext,
  useRef,
  useState,
} from 'react';

import { ACTION, PointerLockProps } from '../common';

const PointerLockContext = createContext<PointerLockProps>(
  {} as PointerLockProps,
);

/**
 *  Pointer lock provider context hook
 *  @returns [PointerLockProps]{@link PointerLockProps}
 */
export const usePointerLock = (): PointerLockProps =>
  useContext(PointerLockContext);

/**
 *  This provides used pointer-lock browser api
 *  It using with all timeblock moving actions
 */
const PointerLockProvider: FC = ({ children }) => {
  const moveRef = useRef<HTMLDivElement | null>(null);
  const resTopRef = useRef<HTMLDivElement | null>(null);
  const resBotRef = useRef<HTMLDivElement | null>(null);

  /**
   * Cursor locking state
   */
  const [isLocking, setIsLocking] = useState<boolean>(false);

  /**
   * Current movement state
   */
  const [movement, setMovement] = useState<number>(0);

  /**
   * Action code
   */
  const [action, setAction] = useState<number>(ACTION.MOVE);

  /**
   * Cursor locking callback
   */
  const startDrawing = useCallback((action: number) => {
    if (!moveRef.current || !resTopRef.current || !resBotRef.current) return;
    setAction(action);
    setIsLocking(true);

    switch (action) {
      case ACTION.MOVE: {
        moveRef.current.requestPointerLock();
        break;
      }
      case ACTION.RESIZE_BOT: {
        resBotRef.current.requestPointerLock();
        break;
      }
      case ACTION.RESIZE_TOP: {
        resTopRef.current.requestPointerLock();
        break;
      }
      default:
        setIsLocking(false);
        break;
    }
  }, []);

  /**
   * Cursor unlocking callback
   */
  const finishDrawing = useCallback(() => {
    setIsLocking(false);
    document.exitPointerLock();
  }, []);

  /**
   * Save current movement
   */
  const draw = useCallback(
    ({ nativeEvent }: React.MouseEvent) => {
      if (!isLocking) {
        return;
      }
      const { movementY } = nativeEvent;
      setMovement(movementY);
    },
    [isLocking],
  );

  return (
    <PointerLockContext.Provider
      value={{
        draw,
        moveRef,
        resBotRef,
        resTopRef,
        isLocking,
        movement,
        startDrawing,
        finishDrawing,
        action,
      }}
    >
      {children}
    </PointerLockContext.Provider>
  );
};

export default PointerLockProvider;
