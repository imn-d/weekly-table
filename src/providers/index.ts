export * from './CanvasProvider';
export * from './CellsProvider';
export * from './CustomTimeProvider';
export * from './PointerLockProvider';
export * from './PopupProvider';
export * from './Scheduler';
export * from './TimeBlockProvider';
