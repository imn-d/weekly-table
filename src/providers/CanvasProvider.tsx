import React, {
  FC,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useRef,
  useState,
} from 'react';

import { CanvasProps } from '../common';
import { useCells } from './CellsProvider';
import { useScheduler } from './Scheduler';

const CanvasContext = createContext<CanvasProps>({} as CanvasProps);
/**
 *  Drawer provider context hook
 *  @returns [CanvasProps]{@link CanvasProps}
 */
export const useCanvas = (): CanvasProps => useContext(CanvasContext);

/**
 * Drawing provider, used for timeblocks creation
 */
const CanvasProvider: FC = ({ children }) => {
  const { headerHeightProp, helperWidthProp, bottomHeightProp } =
    useScheduler();
  const { offsetWidth, offsetHeight, cells, setCells } = useCells();

  /**
   *  Drawing state
   */
  const [isDrawing, setIsDrawing] = useState<boolean>(false);

  /**
   * Canvas reference
   */
  const canvasRef = useRef<HTMLCanvasElement>(null);

  /**
   * Context reference
   */
  const contextRef = useRef<CanvasRenderingContext2D | null>(null);

  /**
   * Start point relative coordinates
   */
  const [start, setStart] = useState<[number, number]>([0, 0]);

  /**
   * End point relative coordinates
   */
  const [end, setEnd] = useState<[number, number] | undefined>(undefined);

  /**
   * Current position relative coordinates
   */
  const [current, setCurrent] = useState<[number, number] | undefined>(
    undefined,
  );

  /**
   * Drawing area init
   */
  useEffect(() => {
    const canvas = canvasRef.current;
    if (!canvas) return;
    canvas.width = offsetWidth * 2;
    canvas.height = offsetHeight * 2;
    canvas.style.width = `${offsetWidth}px`;
    canvas.style.height = `${offsetHeight}px`;
    canvas.draggable = false;

    const context = canvas.getContext('2d');
    if (!context) return;
    context.scale(2, 2);
    context.lineCap = 'round';
    context.strokeStyle = 'blue';
    context.lineWidth = 1;
    context.fillStyle = '#0000FF19';
    contextRef.current = context;
  }, [offsetHeight, offsetWidth]);

  /**
   * Start drawing callback
   */
  const startDrawing = useCallback(({ nativeEvent }: React.MouseEvent) => {
    const { offsetX, offsetY } = nativeEvent;
    if (!contextRef.current) return;
    contextRef.current.beginPath();
    contextRef.current.moveTo(offsetX, offsetY);
    setIsDrawing(true);
    setStart([offsetX, offsetY]);
    setEnd(undefined);
  }, []);

  /**
   * Clearing canvas callback
   */
  const clearCanvas = useCallback(() => {
    if (!contextRef.current) {
      return;
    }
    contextRef.current.clearRect(0, 0, offsetWidth, offsetHeight);
  }, [offsetHeight, offsetWidth]);

  /**
   * End drawing callback
   */
  const finishDrawing = useCallback(
    ({ nativeEvent }: React.MouseEvent) => {
      if (!contextRef.current) return;
      if (!isDrawing) return;

      contextRef.current.closePath();
      setIsDrawing(false);
      setEnd([nativeEvent.offsetX, nativeEvent.offsetY]);
      setCurrent(undefined);
      clearCanvas();
    },
    //eslint-disable-next-line
    [isDrawing],
  );

  /**
   * Drawing process callback
   */
  const draw = useCallback(
    ({ nativeEvent }: React.MouseEvent) => {
      if (!isDrawing || !contextRef.current) {
        return;
      }
      setCurrent([nativeEvent.offsetX, nativeEvent.offsetY]);
      contextRef.current.clearRect(0, 0, offsetWidth, offsetHeight);
      contextRef.current.strokeRect(
        start[0],
        start[1],
        nativeEvent.offsetX - start[0],
        nativeEvent.offsetY - start[1],
      );
      contextRef.current.fillRect(
        start[0] + 1,
        start[1] + 1,
        nativeEvent.offsetX - start[0] - 1,
        nativeEvent.offsetY - start[1] - 1,
      );
    },
    [isDrawing, offsetHeight, offsetWidth, start],
  );

  /**
   * Finding cell index by coordinates
   */
  const findCell = useCallback(
    (position: [number, number] | undefined): number => {
      if (!position || !cells) return -1;
      if (position[0] >= offsetWidth - 5) position[0] = offsetWidth - 5;
      if (position[0] <= helperWidthProp) position[0] = helperWidthProp;
      if (position[1] <= headerHeightProp) position[1] = headerHeightProp;
      if (position[1] >= offsetHeight - bottomHeightProp)
        position[1] = offsetHeight - bottomHeightProp - 1;

      return cells.findIndex(
        (cell) =>
          cell.position.x <= position[0] &&
          position[0] <= cell.position.w &&
          cell.position.y <= position[1] &&
          position[1] <= cell.position.h,
      );
    },
    [
      offsetWidth,
      cells,
      helperWidthProp,
      headerHeightProp,
      offsetHeight,
      bottomHeightProp,
    ],
  );

  /**
   * Start, end, current coordinates changes watcher
   * Calculates selected cells area and set isSelected property to it
   */
  useEffect(() => {
    if (end === undefined && current === undefined) return;
    const startCell = findCell(start);
    if (startCell === -1) return;
    const endCell = findCell(current === undefined ? end : current);
    if (endCell === -1) return;
    const sCell = cells[startCell];
    const eCell = cells[endCell];
    const sRow = sCell.row < eCell.row ? sCell.row : eCell.row;
    const sCol = sCell.column < eCell.column ? sCell.column : eCell.column;
    const eRow = eCell.row > sCell.row ? eCell.row : sCell.row;
    const eCol = eCell.column > sCell.column ? eCell.column : sCell.column;

    const insertCells = cells.map((cell) => {
      if (
        cell.row >= sRow &&
        cell.row <= eRow &&
        cell.column >= sCol &&
        cell.column <= eCol
      ) {
        return { ...cell, isSelected: true };
      } else return cell;
    });
    setCells(insertCells);
    //eslint-disable-next-line
  }, [start, end, current]);

  return (
    <CanvasContext.Provider
      value={{
        canvasRef,
        startDrawing,
        finishDrawing,
        draw,
        isDrawing,
      }}
    >
      {children}
    </CanvasContext.Provider>
  );
};

export default CanvasProvider;
