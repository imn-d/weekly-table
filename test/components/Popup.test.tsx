import { mount } from 'enzyme';
import React from 'react';

import {
  ACTION,
  CanvasProps,
  CellsProps,
  CustomTimeProps,
  PointerLockProps,
  PopupProps,
  PopupType,
  SchedulerProps,
  TimeBlock,
  TimeBlockProps,
} from '../../src/common';
import { Popup } from '../../src/components';
import * as ucn from '../../src/providers/CanvasProvider';
import * as uc from '../../src/providers/CellsProvider';
import * as uct from '../../src/providers/CustomTimeProvider';
import * as upl from '../../src/providers/PointerLockProvider';
import * as up from '../../src/providers/PopupProvider';
import * as us from '../../src/providers/Scheduler';
import * as ut from '../../src/providers/TimeBlockProvider';

describe('Popup Component', () => {
  const useSchedulerMock = jest.spyOn(us, 'useScheduler');
  const useCellMock = jest.spyOn(uc, 'useCells');
  const useTimeBlockMock = jest.spyOn(ut, 'useTimeBlock');
  const usePopupMock = jest.spyOn(up, 'usePopup');
  const useCustomMock = jest.spyOn(uct, 'useCustomTime');
  const useCanvasMock = jest.spyOn(ucn, 'useCanvas');
  const useLockMock = jest.spyOn(upl, 'usePointerLock');

  const setTopPositionMock = jest.fn();
  const setBotPositionMock = jest.fn();

  const showPopup = jest.fn();
  const hidePopup = jest.fn();
  const deleteBlock = jest.fn();

  const setEditing = jest.fn();
  const updateBlock = jest.fn();
  const resetTime = jest.fn();

  const startDrawing = jest.fn();
  const finishDrawing = jest.fn();
  const draw = jest.fn();

  const timeBlock = {
    id: '1',
    isTemp: false,
    realStartTime: 100000,
    realEndTime: 200000,
    startTime: 100000,
    endTime: 200000,
    column: 1,
    top: 0,
    height: 100,
    left: 20,
    width: 50,
  } as TimeBlock;

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const popupMock = {
    popup: { id: '1', block: timeBlock } as PopupType,
    showPopup,
    hidePopup,
    deleteBlock,
    isAllowing: true,
  } as PopupProps;

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const customMock = {
    isEditing: false,
    setEditing,
    updateBlock,
    resetTime,
  } as CustomTimeProps;

  const canvasMock = {
    isDrawing: false,
  } as CanvasProps;

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const lockMock = {
    moveRef: { current: null },
    resBotRef: { current: null },
    resTopRef: { current: null },
    startDrawing,
    finishDrawing,
    draw,
    isLocking: false,
  } as PointerLockProps;

  beforeEach(() => {
    jest.clearAllMocks();

    useSchedulerMock.mockReturnValue({ baseZIndex: 0 } as SchedulerProps);
    useCellMock.mockReturnValue({
      widthStep: 50,
      heightStep: 50,
    } as CellsProps);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    useTimeBlockMock.mockReturnValue({
      setTopPosition: setTopPositionMock,
      setBotPosition: setBotPositionMock,
    } as TimeBlockProps);
  });

  describe('Base Operations', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      usePopupMock.mockReturnValue(popupMock);
      useCustomMock.mockReturnValue(customMock);
      useCanvasMock.mockReturnValue(canvasMock);
      useLockMock.mockReturnValue(lockMock);
    });

    it('Base Render', () => {
      const wrapper = mount(<Popup />);

      expect(useSchedulerMock).toBeCalledTimes(1);
      expect(useCellMock).toBeCalledTimes(1);
      expect(useTimeBlockMock).toBeCalledTimes(1);
      expect(usePopupMock).toBeCalledTimes(1);
      expect(useCustomMock).toBeCalledTimes(1);
      expect(useCanvasMock).toBeCalledTimes(1);
      expect(useLockMock).toBeCalledTimes(1);

      expect(wrapper).toMatchSnapshot('baseRender');
      wrapper.unmount();
    });

    it('Wrapper Actions Default', () => {
      const wrapper = mount(<Popup />);
      const parentDiv = wrapper.find('#popupWrapper').hostNodes();

      expect(parentDiv).toHaveLength(1);
      expect(showPopup).toBeCalledTimes(0);
      expect(hidePopup).toBeCalledTimes(0);
      expect(finishDrawing).toBeCalledTimes(0);

      parentDiv.simulate('mouseenter');
      parentDiv.simulate('mousemove');
      expect(showPopup).toBeCalledTimes(1);
      expect(showPopup).toBeCalledWith('1');

      parentDiv.simulate('mouseleave');
      expect(hidePopup).toBeCalledTimes(1);
      expect(finishDrawing).toBeCalledTimes(1);

      wrapper.unmount();
    });

    it('Popup Top Default', () => {
      const wrapper = mount(<Popup />);
      const parentDiv = wrapper.find('#popupTop').hostNodes();

      expect(parentDiv).toHaveLength(1);
      expect(setTopPositionMock).toBeCalledTimes(0);
      expect(startDrawing).toBeCalledTimes(0);
      expect(finishDrawing).toBeCalledTimes(0);
      expect(draw).toBeCalledTimes(0);

      parentDiv.simulate('mousemove');
      expect(draw).toBeCalledTimes(1);

      parentDiv.simulate('mousedown');
      expect(startDrawing).toBeCalledTimes(1);
      expect(startDrawing).toBeCalledWith(ACTION.RESIZE_TOP);

      parentDiv.simulate('mouseup');
      expect(finishDrawing).toBeCalledTimes(1);
      expect(setTopPositionMock).toBeCalledTimes(1);
      expect(setTopPositionMock).toBeCalledWith(timeBlock, true);

      wrapper.unmount();
    });

    it('Popup Center Default', () => {
      const wrapper = mount(<Popup />);
      const parentDiv = wrapper.find('#popupCenter').hostNodes();

      expect(parentDiv).toHaveLength(1);
      expect(setTopPositionMock).toBeCalledTimes(0);
      expect(startDrawing).toBeCalledTimes(0);
      expect(finishDrawing).toBeCalledTimes(0);
      expect(draw).toBeCalledTimes(0);

      parentDiv.simulate('mousemove');
      expect(draw).toBeCalledTimes(1);

      parentDiv.simulate('mousedown');
      expect(startDrawing).toBeCalledTimes(1);
      expect(startDrawing).toBeCalledWith(ACTION.MOVE);

      parentDiv.simulate('mouseup');
      expect(finishDrawing).toBeCalledTimes(1);
      expect(setTopPositionMock).toBeCalledTimes(1);
      expect(setTopPositionMock).toBeCalledWith(timeBlock, false);

      wrapper.unmount();
    });

    it('Popup Bot Default', () => {
      const wrapper = mount(<Popup />);
      const parentDiv = wrapper.find('#popupBot').hostNodes();

      expect(parentDiv).toHaveLength(1);
      expect(setBotPositionMock).toBeCalledTimes(0);
      expect(startDrawing).toBeCalledTimes(0);
      expect(finishDrawing).toBeCalledTimes(0);
      expect(draw).toBeCalledTimes(0);

      parentDiv.simulate('mousemove');
      expect(draw).toBeCalledTimes(1);

      parentDiv.simulate('mousedown');
      expect(startDrawing).toBeCalledTimes(1);
      expect(startDrawing).toBeCalledWith(ACTION.RESIZE_BOT);

      parentDiv.simulate('mouseup');
      expect(finishDrawing).toBeCalledTimes(1);
      expect(setBotPositionMock).toBeCalledTimes(1);
      expect(setBotPositionMock).toBeCalledWith(timeBlock);

      wrapper.unmount();
    });
  });

  it('Null Render', () => {
    usePopupMock.mockReturnValueOnce({ ...popupMock, popup: undefined });
    useCustomMock.mockReturnValueOnce(customMock);
    useCanvasMock.mockReturnValueOnce(canvasMock);
    useLockMock.mockReturnValueOnce(lockMock);

    const wrapper = mount(<Popup />);
    expect(wrapper).toMatchInlineSnapshot(`<Popup />`);
    wrapper.unmount();
  });

  it('Wrapper Actions True', () => {
    usePopupMock.mockReturnValueOnce(popupMock);
    useCustomMock.mockReturnValueOnce({ ...customMock, isEditing: true });
    useCanvasMock.mockReturnValueOnce(canvasMock);
    useLockMock.mockReturnValueOnce({ ...lockMock, isLocking: true });

    const wrapper = mount(<Popup />);
    const parentDiv = wrapper.find('#popupWrapper').hostNodes();

    expect(parentDiv).toHaveLength(1);

    parentDiv.simulate('mouseenter');
    expect(showPopup).toBeCalledWith('1');

    parentDiv.simulate('mousemove');
    parentDiv.simulate('mouseleave');

    expect(showPopup).toBeCalledTimes(2);
    expect(hidePopup).toBeCalledTimes(0);
    expect(finishDrawing).toBeCalledTimes(0);

    wrapper.unmount();
  });

  describe('Popup Display Tests', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      useCanvasMock.mockReturnValue(canvasMock);
      useLockMock.mockReturnValue(lockMock);
    });

    it('Popup Top Display', () => {
      usePopupMock.mockReturnValueOnce(popupMock);
      useCustomMock.mockReturnValueOnce(customMock);

      let wrapper = mount(<Popup />);
      let parentDiv = wrapper.find('#popupTop').hostNodes();
      expect(parentDiv.props()?.style?.display).toBe('flex');
      wrapper.unmount();

      usePopupMock.mockReturnValueOnce(popupMock);
      useCustomMock.mockReturnValueOnce({ ...customMock, isEditing: true });

      wrapper = mount(<Popup />);
      parentDiv = wrapper.find('#popupTop').hostNodes();
      expect(parentDiv.props()?.style?.display).toBe('none');
      wrapper.unmount();

      usePopupMock.mockReturnValueOnce({ ...popupMock, isAllowing: false });
      useCustomMock.mockReturnValueOnce(customMock);

      wrapper = mount(<Popup />);
      parentDiv = wrapper.find('#popupTop').hostNodes();
      expect(parentDiv.props()?.style?.display).toBe('none');
      wrapper.unmount();
    });

    it('Popup Center Display', () => {
      usePopupMock.mockReturnValueOnce(popupMock);
      useCustomMock.mockReturnValueOnce(customMock);

      let wrapper = mount(<Popup />);
      let parentDiv = wrapper.find('#popupCenter').hostNodes();
      expect(parentDiv.props()?.style?.display).toBe('block');
      wrapper.unmount();

      usePopupMock.mockReturnValueOnce(popupMock);
      useCustomMock.mockReturnValueOnce({ ...customMock, isEditing: true });

      wrapper = mount(<Popup />);
      parentDiv = wrapper.find('#popupCenter').hostNodes();
      expect(parentDiv.props()?.style?.display).toBe('none');
      wrapper.unmount();

      usePopupMock.mockReturnValueOnce({ ...popupMock, isAllowing: false });
      useCustomMock.mockReturnValueOnce(customMock);

      wrapper = mount(<Popup />);
      parentDiv = wrapper.find('#popupCenter').hostNodes();
      expect(parentDiv.props()?.style?.display).toBe('none');
      wrapper.unmount();
    });

    it('Popup Bot Display', () => {
      usePopupMock.mockReturnValueOnce(popupMock);
      useCustomMock.mockReturnValueOnce(customMock);

      let wrapper = mount(<Popup />);
      let parentDiv = wrapper.find('#popupBot').hostNodes();
      expect(parentDiv.props()?.style?.display).toBe('flex');
      wrapper.unmount();

      usePopupMock.mockReturnValueOnce(popupMock);
      useCustomMock.mockReturnValueOnce({ ...customMock, isEditing: true });

      wrapper = mount(<Popup />);
      parentDiv = wrapper.find('#popupBot').hostNodes();
      expect(parentDiv.props()?.style?.display).toBe('none');
      wrapper.unmount();

      usePopupMock.mockReturnValueOnce({ ...popupMock, isAllowing: false });
      useCustomMock.mockReturnValueOnce(customMock);

      wrapper = mount(<Popup />);
      parentDiv = wrapper.find('#popupBot').hostNodes();
      expect(parentDiv.props()?.style?.display).toBe('none');
      wrapper.unmount();
    });
  });

  describe('Popup Right Tests', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      usePopupMock.mockReturnValue(popupMock);
      useCanvasMock.mockReturnValue(canvasMock);
    });

    it('Popup Right Display', () => {
      useCustomMock.mockReturnValue(customMock);
      useLockMock.mockReturnValueOnce(lockMock);

      let wrapper = mount(<Popup />);
      let parentDiv = wrapper.find('#popupRight').hostNodes();
      expect(parentDiv.props()?.style?.display).toBe('flex');
      wrapper.unmount();

      useLockMock.mockReturnValueOnce({ ...lockMock, isLocking: true });

      wrapper = mount(<Popup />);
      parentDiv = wrapper.find('#popupRight').hostNodes();
      expect(parentDiv.props()?.style?.display).toBe('none');
      wrapper.unmount();
    });

    it('Popup Right isEditing Test', () => {
      useCustomMock.mockReturnValueOnce(customMock);
      useLockMock.mockReturnValueOnce(lockMock);

      let wrapper = mount(<Popup />);
      let parentDiv = wrapper.find('#popupCheckIcon').hostNodes();
      expect(parentDiv).toHaveLength(0);
      wrapper.unmount();

      useCustomMock.mockReturnValueOnce({
        ...customMock,
        isEditing: true,
        updateBlock: jest
          .fn()
          .mockReturnValueOnce(false)
          .mockReturnValueOnce(true),
      });

      wrapper = mount(<Popup />);
      parentDiv = wrapper.find('#popupCheckIcon').hostNodes();
      expect(parentDiv).toHaveLength(1);

      expect(setEditing).toBeCalledTimes(0);
      parentDiv.simulate('click');
      expect(setEditing).toBeCalledTimes(1);
      expect(setEditing).toBeCalledWith(false);
      setEditing.mockClear();

      parentDiv.simulate('click');
      expect(setEditing).toBeCalledTimes(1);
      expect(setEditing).toBeCalledWith(true);

      wrapper.unmount();
    });

    it('Popup Right CEIcon Test', () => {
      useLockMock.mockReturnValueOnce(lockMock);
      useCustomMock.mockReturnValueOnce(customMock);

      const wrapper = mount(<Popup />);
      const parentDiv = wrapper.find('#popupCEIcon').hostNodes();
      expect(parentDiv).toHaveLength(1);

      expect(resetTime).toBeCalledTimes(0);
      expect(setEditing).toBeCalledTimes(0);

      parentDiv.simulate('click');
      expect(resetTime).toBeCalledTimes(1);
      expect(setEditing).toBeCalledTimes(1);
      wrapper.unmount();
    });

    it('Popup Right DeleteIcon Test', () => {
      useLockMock.mockReturnValueOnce(lockMock);
      useCustomMock.mockReturnValueOnce(customMock);

      const wrapper = mount(<Popup />);
      const parentDiv = wrapper.find('#popupDeleteIcon').hostNodes();
      expect(parentDiv).toHaveLength(1);

      expect(deleteBlock).toBeCalledTimes(0);

      parentDiv.simulate('click');
      expect(deleteBlock).toBeCalledTimes(1);
      expect(deleteBlock).toBeCalledWith('1');
      wrapper.unmount();
    });

    it('Popup Right DeleteIcon Visibility Test', () => {
      useLockMock.mockReturnValue(lockMock);
      useCustomMock.mockReturnValueOnce({ ...customMock, isEditing: false });

      let wrapper = mount(<Popup />);
      let parentDiv = wrapper.find('#popupDeleteIcon').hostNodes();
      expect(parentDiv.props()?.style?.display).toBe('block');
      wrapper.unmount();

      useCustomMock.mockReturnValueOnce({ ...customMock, isEditing: true });

      wrapper = mount(<Popup />);
      parentDiv = wrapper.find('#popupDeleteIcon').hostNodes();
      expect(parentDiv.props()?.style?.display).toBe('none');
      wrapper.unmount();
    });
  });
});
