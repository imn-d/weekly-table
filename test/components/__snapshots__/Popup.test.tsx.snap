// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup Component Base Operations Base Render: baseRender 1`] = `
<Popup>
  <div
    className="scheduler-popup-wrapper"
    id="popupWrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    style={
      Object {
        "boxSizing": "border-box",
        "display": "grid",
        "gridTemplateColumns": "15px 37.5px 30px",
        "gridTemplateRows": "33px 103px 33px ",
        "left": 5,
        "maxWidth": 82.5,
        "position": "absolute",
        "top": -35,
        "width": 82.5,
        "zIndex": 5,
      }
    }
  >
    <div
      style={
        Object {
          "gridColumn": " 1",
          "gridRow": "1 / 4",
        }
      }
    />
    <div
      className="scheduler-popup scheduler-popup-top"
      id="popupTop"
      onMouseDown={[Function]}
      onMouseMove={[MockFunction]}
      onMouseUp={[Function]}
      style={
        Object {
          "alignItems": "self-end",
          "cursor": "n-resize",
          "display": "flex",
          "gridColumn": "2",
          "gridRow": "1",
          "justifyContent": "center",
        }
      }
    >
      <Memo()>
        <svg
          className="scheduler-icon drag-icon"
          height={25}
          id="dragIcon"
          viewBox="0 0 25 25"
          width={25}
          x="0px"
          xmlns="http://www.w3.org/2000/svg"
          y="0px"
        >
          <path
            d="M20 9H4v2h16V9zM4 15h16v-2H4v2z"
          />
        </svg>
      </Memo()>
    </div>
    <div
      className="scheduler-popup scheduler-popup-right"
      id="popupRight"
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "column",
          "gridColumn": " 3",
          "gridRow": "1 / 4",
          "justifyContent": "space-around",
        }
      }
    >
      <div
        className="popup-cancel-edit-icon"
        id="popupCEIcon"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <Memo()>
          <svg
            className="scheduler-icon edit-icon"
            height={25}
            id="editIcon"
            viewBox="0 0 25 25"
            width={25}
            x="0px"
            xmlns="http://www.w3.org/2000/svg"
            y="0px"
          >
            <path
              d="M 18.414062 2 C 18.158188 2 17.902031 2.0974687 17.707031 2.2929688 L 16 4 L 20 8 L 21.707031 6.2929688 C 22.098031 5.9019687 22.098031 5.2689063 21.707031 4.8789062 L 19.121094 2.2929688 C 18.925594 2.0974687 18.669937 2 18.414062 2 z M 14.5 5.5 L 3 17 L 3 21 L 7 21 L 18.5 9.5 L 14.5 5.5 z"
            />
          </svg>
        </Memo()>
      </div>
      <div
        className="popup-delete-icon"
        id="popupDeleteIcon"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
            "display": "block",
          }
        }
      >
        <Memo()>
          <svg
            className="scheduler-icon delete-icon"
            height={25}
            id="deleteIcon"
            viewBox="0 0 30 30"
            width={25}
            x="0px"
            xmlns="http://www.w3.org/2000/svg"
            y="0px"
          >
            <path
              d="M 13 3 A 1.0001 1.0001 0 0 0 11.986328 4 L 6 4 A 1.0001 1.0001 0 1 0 6 6 L 24 6 A 1.0001 1.0001 0 1 0 24 4 L 18.013672 4 A 1.0001 1.0001 0 0 0 17 3 L 13 3 z M 6 8 L 6 24 C 6 25.105 6.895 26 8 26 L 22 26 C 23.105 26 24 25.105 24 24 L 24 8 L 6 8 z"
            />
          </svg>
        </Memo()>
      </div>
    </div>
    <div
      className="scheduler-popup scheduler-popup-center"
      id="popupCenter"
      onMouseDown={[Function]}
      onMouseMove={[MockFunction]}
      onMouseUp={[Function]}
      style={
        Object {
          "cursor": "grab",
          "display": "block",
          "gridColumn": "2",
          "gridRow": "2",
        }
      }
    />
    <div
      className="scheduler-popup scheduler-popup-bot"
      id="popupBot"
      onMouseDown={[Function]}
      onMouseMove={[MockFunction]}
      onMouseUp={[Function]}
      style={
        Object {
          "alignItems": "self-start",
          "cursor": "n-resize",
          "display": "flex",
          "gridColumn": "2",
          "gridRow": "3",
          "justifyContent": "center",
        }
      }
    >
      <Memo()>
        <svg
          className="scheduler-icon drag-icon"
          height={25}
          id="dragIcon"
          viewBox="0 0 25 25"
          width={25}
          x="0px"
          xmlns="http://www.w3.org/2000/svg"
          y="0px"
        >
          <path
            d="M20 9H4v2h16V9zM4 15h16v-2H4v2z"
          />
        </svg>
      </Memo()>
    </div>
  </div>
</Popup>
`;
